sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala

```diff
@@ -92,6 +94,30 @@ class AdaptiveQueryExecSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("Change merge join to broadcast join and reduce number of shuffle partitions") {
```

Comment:
@maryannxue, added an end2end test where SMJ->BHJ and reduce-shuffle-num happen at the same time .  Actually `REDUCE_POST_SHUFFLE_PARTITIONS_ENABLED` is enabled by default, so other test cases also reduce shuffle partition numbers if we don't disable it. But that doesn't seem to be a problem.